### DISCUSSION TITLE
What is the simplest way to run pkg-config commands in a dune file [question]?

### DISCUSSION THREAD

**Discussion by UnixJunkie** [ocaml/dune#9070] [Q&A]
Hello,

To compile some C bindings, I would like dune to use the output of some pkg-config commands, such as:
- pkg-config --libs nlopt
- pkg-config --cflags nlopt

Am I dreaming, or dune has not support for creating a variable whose content is the output
of a system command (like the back quoting mechanism in the bash shell) ?

Thanks,
F.

**Comment by UnixJunkie** [comment]
If this is not supported, then I would strongly suggest it as a feature request.

**Comment by NielsMommen** [comment]
You can use dune's [`Configurator`](https://dune.readthedocs.io/en/stable/dune-libs.html) library for this purpose. It has a [`Pkg_config`](https://github.com/ocaml/dune/blob/2f0a93d22b80017aee5cd73855b940f17a19ef56/otherlibs/configurator/src/v1.mli#L72C1-L107C30) module which allows to [query](https://github.com/ocaml/dune/blob/2f0a93d22b80017aee5cd73855b940f17a19ef56/otherlibs/configurator/src/v1.mli#L95C1-L105C39) information about a package through pkg-config. You can then write the queried flags to sexpression files and read those generated files when compiling the C bindings.

# Example
Consider the case where we want to link to some C code that uses `gtksourceview2`. Therefore, we would like to query flags through pkg-config in order to link the foreign stubs.

Assume you have following file structure
```
|-- lib
|   |-- flags
|   |   |-- discover.ml
|   |   |-- dune
|   |--
|   |-- foo_c.c
|   |-- foo.ml
|   |-- dune
|--
```

The dune file in `lib/flags` has following content:
```dune
(executable
 (name discover)
 (libraries dune-configurator))

(rule
 (targets cflags.sexp libs.sexp)
 (action
  (run %{dep:discover.exe})))
```
It declares an executable `discover` that generates `cflags.sexp` and `libs.sexp`. These files contain sexpressions which contain the flags discovered for `gtksourceview2` through pkg-config. The implementation of `discover` could be something like:
```ocaml
module C = Configurator.V1

let get_gtksourceview_pkg_conf c =
  match C.Pkg_config.get c with
  | None -> C.die "Could not find 'pkg-config'"
  | Some pkg_conf -> (
      match
        C.Pkg_config.query_expr_err pkg_conf ~package:"gtksourceview-2.0"
          ~expr:""
      with
      | Ok cfg -> cfg
      | Error err -> C.die "%s" err)

let () =
  C.main ~args:[] ~name:"discover_gtksourceview_flags" @@ fun c ->
  let C.Pkg_config.{ cflags; libs } = get_gtksourceview_pkg_conf c in
  C.Flags.write_sexp "cflags.sexp" cflags;
  C.Flags.write_sexp "libs.sexp" libs
```

The user rule simply runs `discover` in order to generate `cflags.sexp` and `libs.sexp`

The dune file in `lib` looks like:
```dune
(library
 (name foo)
 (libraries lablgtk2 lablgtk2.sourceview2)
 ; Pass flags to c compiler when constructing the library
 ; archive file for the C stubs. E.g. to pass -lbar to the linker.
 (c_library_flags
  (:standard
   (:include flags/libs.sexp)))
 (foreign_stubs
  (language c)
  (flags
   (:include flags/cflags.sexp))
  (names foo_c)))
```
We use `:include` forms in order to include the flags that we previously generated. There is [another example](https://dune.readthedocs.io/en/latest/quick-start.html#defining-a-library-with-c-stubs-using-pkg-config) in dune's documentation.

**Comment by UnixJunkie** [comment]
this is too ugly

**Comment by UnixJunkie** [comment]
we should just specify in the dune file directly what is the command to run in order to get those flags

**Comment by Khady** [comment]
> this is too ugly

What a interesting argument...

More seriously, nothing blocks you from having the pkg-config call directly in dune. There aren't variables, but the output can be put into a file which is used by another rule.

**Reply by UnixJunkie** [reply]
this is so ugly that it reminds me the ocamlbuild plugins that some people were so fan of writing (yes, code to compile your code); I never wrote a single one in my life
